pymc/smc/smc.py

```diff
@@ -384,34 +364,43 @@ def tune(self):
 
     def mutate(self):
         """Independent Metropolis-Hastings perturbation."""
-        ac_ = np.empty((self.n_steps, self.draws))
-        log_R = np.log(self.rng.random((self.n_steps, self.draws)))
-
-        # The proposal is independent from the current point.
-        # We have to take that into account to compute the Metropolis-Hastings acceptance
-        # We first compute the logp of proposing a transition to the current points.
-        # This variable is updated at the end of the loop with the entries from the accepted
-        # transitions, which is equivalent to recomputing it in every iteration of the loop.
-        backward_logp = self.proposal_dist.logpdf(self.tempered_posterior)
-        for n_step in range(self.n_steps):
+        self.n_steps = 1
+        old_corr = 2
+        corr = Pearson(self.tempered_posterior)
+        ac_ = []
+        while True:
+            log_R = np.log(self.rng.random(self.draws))
+            # The proposal is independent from the current point.
+            # We have to take that into account to compute the Metropolis-Hastings acceptance
+            # We first compute the logp of proposing a transition to the current points.
+            # This variable is updated at the end of the loop with the entries from the accepted
+            # transitions, which is equivalent to recomputing it in every iteration of the loop.
             proposal = floatX(self.proposal_dist.rvs(size=self.draws, random_state=self.rng))
             proposal = proposal.reshape(len(proposal), -1)
-            # We then compute the logp of proposing a transition to the new points
+            # To do that we compute the logp of moving to a new point
             forward_logp = self.proposal_dist.logpdf(proposal)
-
+            # And to going back from that new point
+            backward_logp = self.proposal_dist.logpdf(self.tempered_posterior)
             ll = np.array([self.likelihood_logp_func(prop) for prop in proposal])
             pl = np.array([self.prior_logp_func(prop) for prop in proposal])
             proposal_logp = pl + ll * self.beta
-            accepted = log_R[n_step] < (
+            accepted = log_R < (
                 (proposal_logp + backward_logp) - (self.tempered_posterior_logp + forward_logp)
             )
 
-            ac_[n_step] = accepted
-            self.tempered_posterior[accepted] = proposal[accepted]
-            self.tempered_posterior_logp[accepted] = proposal_logp[accepted]
-            self.prior_logp[accepted] = pl[accepted]
-            self.likelihood_logp[accepted] = ll[accepted]
-            backward_logp[accepted] = forward_logp[accepted]
+            tmp = np.copy(self.tempered_posterior)
```

Comment:
why do we need to copy the tempered posterior?

Reply:
in case we reject the autocorrelation condition on line 394

Reply:
Why not just do
```python
pearson_r = corr.get(proponal[accepted])
```

Or do you think the repeated indexing is slower than the copy operation?

Reply:
In that case you will be evaluating only the new (accepted) positions, but you need to evaluate both the new and old, i.e. those that you did not change because you rejected the proposal.

Reply:
practicality beats purity. I like it!